watertap/tools/analysis_tools/loop_tool/loop_tool.py

```diff
@@ -0,0 +1,732 @@
+            "update_sweep_params_before_init", False
+        )
+        self.build_outputs_kwargs = self.options.get("build_outputs_kwargs", None)
+        # generated combined build kwargs (default + loop)
+        self.combined_build_defaults = {}  # self.build_default
+        self.combined_build_defaults.update(self.options.get("build_defaults", {}))
+        self.combined_build_defaults.update(self.build_default)
+        # generated combined optimize kwargs (default + loop)
+        self.combined_optimize_defaults = {}
+        self.combined_optimize_defaults.update(
+            self.options.get("optimize_defaults", {})
+        )
+        self.combined_optimize_defaults.update(self.optimize_defaults)
+        # generated combined init  kwargs (default + loop)
+        self.combined_init_defaults = {}
+        self.combined_init_defaults.update(self.options.get("init_defaults", {}))
+        self.combined_init_defaults.update(self.init_defaults)
+
+    def _check_solution_exists(self):
+        """hidden function to check if solution
+        exists, and create and h5 file for data storage"""
+        self.h5_backup_location = create_backup_file(
+            self.h5_file_location, self.h5_backup_location, self.h5_directory
+        )
+        create_h5_file(self.h5_file_location)
+
+        sucess_feasible = self.expected_num_samples
+        run_sweep = merge_data_into_file(
+            self.h5_file_location,
+            self.h5_backup_location,
+            self.h5_directory,
+            expected_solved_values=self.expected_num_samples,
+            min_solve_values=self.min_num_samples,
+            force_rerun=self.force_rerun,
+        )
+        return run_sweep
+
+    def check_solution_exists(self):
+        """check if solution exists, ensuring to use only
+        rank 0 if user running MPI, otherwise do direct
+        solution check"""
+        self.cur_h5_file = (self.h5_file_location, self.h5_directory)
+        if has_mpi_peer_processes():
+            mpi_comm = get_mpi_comm_process()
+            results = np.empty(mpi_comm.Get_size(), dtype=bool)
+
+            results[:] = True
+            if mpi_comm.Get_rank() == 0:
+                success = self._check_solution_exists()
+                results[:] = success
+
+            mpi_comm.Bcast(results, root=0)
+            success = results[mpi_comm.Get_rank()]
+        else:
+            success = self._check_solution_exists()
+        return success
+
+    def _create_save_directory(self, save_dir):
+        """used to create a save directory (outputs)"""
+        try:
+            os.mkdir(save_dir)
+        except OSError as error:
+            pass
+
+    def _default_build_ouptuts(self, model, output_keys):
+        outputs = {}
+        for key, pyo_object in output_keys.items():
+            outputs[key] = model.find_component(pyo_object)
+        return outputs
+
+    def run_parameter_sweep(self):
```

Comment:
Instead of constructing the keyword argument dictionary here and constructing the parameter sweep object, can we call the [parameter_sweep](https://github.com/watertap-org/watertap/blob/971dbce4d86bdff720c3bf6f54823f820a0fa687/watertap/tools/parameter_sweep/parameter_sweep_functions.py#L22) function? Same goes for the differential parameter sweep function later in this file

Reply:
I need to use the ps API directly, as some options are not available in parmenter functions (Which IMO should be depreciated, not sure why we even have that). (such as custom_function and custom_function_kwargs).